model/publishing/test/TestBackupService.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2020 (original work) Open Assessment Technologies SA ;
+ */
+declare(strict_types=1);
+
+namespace oat\taoPublishing\model\publishing\test;
+
+use Exception;
+use common_Exception;
+use core_kernel_classes_Resource;
+use oat\generis\model\OntologyAwareTrait;
+use oat\oatbox\filesystem\File;
+use oat\oatbox\service\ConfigurableService;
+use oat\tao\model\import\service\RdsResourceNotFoundException;
+use oat\taoDeliveryRdf\model\DeliveryAssemblyService;
+use oat\taoQtiTest\helpers\QtiPackageExporter;
+
+class TestBackupService extends ConfigurableService
+{
+    use OntologyAwareTrait;
+
+    public const FILESYSTEM_ID = 'publishing';
+
+    public const PROPERTY_QTI_TEST_BACKUP_PATH = 'http://www.tao.lu/Ontologies/TAOPublishing.rdf#QtiTestBackupPath';
+
+    /**
+     * @param string $deliveryUri
+     * @return File
+     * @throws RdsResourceNotFoundException
+     * @throws common_Exception
+     */
+    public function backupDeliveryTestPackage(string $deliveryUri): File
+    {
+        try {
+            $test = $this->getTestResource($deliveryUri);
+            /** @var QtiPackageExporter $packageExporter */
+            $packageExporter = $this->getServiceLocator()->get(QtiPackageExporter::SERVICE_ID);
+
+            return $packageExporter->exportQtiTestPackageToFile($test->getUri(), self::FILESYSTEM_ID, $this->prepareBackupFilePath($deliveryUri));
+        } catch (RdsResourceNotFoundException $e) {
+            $this->logError($e->getMessage());
+            throw $e;
+        } catch (Exception $e) {
+            $this->logError($e->getMessage());
+            throw new common_Exception(sprintf('Backup of origin QTI test failed for delivery %s', $deliveryUri));
+        }
+    }
+
+    /**
+     * @param string $deliveryUri
+     * @return string
+     */
+    private function prepareBackupFilePath(string $deliveryUri): string
+    {
+        $folderName = md5($deliveryUri);
```

Comment:
In the past, we had problems with paths that were not `identifiable` just looking to the folder structure. Are we sure this won't be the same case? :) Also, it's really meant to be overwritten in case of a second call for the same delivery?

Reply:
We can publish can create delivery only once (and we have to export QTI test only when  we create delivery). When delivery is updated it changes just metadata and there is no need to export test again. 
Good point for folder name. I was looking for some method  in TAO we could reuse and couldn't find anything. We also cannot use delivery uri because it contains a lot of special characters. Maybe we could use uri and just remove all special characters, or use unique ID part of uri.